File: src/main/java/net/onelitefeather/pandorascluster/commands/ClaimCommand.java
package net.onelitefeather.pandorascluster.commands;

import cloud.commandframework.annotations.CommandDescription;
import cloud.commandframework.annotations.CommandMethod;
import net.kyori.adventure.text.Component;
import net.onelitefeather.pandorascluster.api.PandorasClusterApi;
import net.onelitefeather.pandorascluster.land.Land;
import net.onelitefeather.pandorascluster.land.player.LandPlayer;
import net.onelitefeather.pandorascluster.util.ChunkUtil;
import org.bukkit.Chunk;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

public record ClaimCommand(PandorasClusterApi api) {

    public ClaimCommand(@NotNull PandorasClusterApi api) {
        this.api = api;
    }

    @CommandMethod("land claim")
    @CommandDescription("Claim a free chunk")
    public void execute(@NotNull Player player) {

        LandPlayer landPlayer = this.api.getLandPlayer(player.getUniqueId());
        if (landPlayer == null) {
            player.sendMessage("Could not find your player data.");
            return;
        }

        Chunk playerChunk = player.getChunk();

        if (this.api.isChunkClaimed(player.getChunk())) {
            player.sendMessage(Component.text("This chunk was already claimed!"));
            return;
        }

        this.api.getLandService().findConnectedChunk(player, land -> {

            var chunkX = playerChunk.getX();
            var chunkZ = playerChunk.getZ();

            Chunk claimedChunk = null;

            for (int x = -2; x < 2 && claimedChunk == null; x++) {
                for (int z = -2; z < 2 && claimedChunk == null; z++) {
                    Chunk chunk = player.getWorld().getChunkAt(x + chunkX, z + chunkZ);
                    if (this.api.isChunkClaimed(chunk)) {
                        claimedChunk = chunk;
                    }
                }
            }

            if (land != null) {

                if (claimedChunk != null) {
                    Land claimedLand = this.api.getLand(claimedChunk);
                    if (claimedLand != null && !ChunkUtil.hasSameOwner(land, claimedLand)) {
                        player.sendMessage(Component.text("distance"));
                        return;
                    }
                }

                if (!land.isOwner(player.getUniqueId())) {
                    player.sendMessage(Component.text("You´re not the Owner from this Land!"));
                    return;
                }

                this.api.getLandService().addChunkPlaceholder(playerChunk, land);
                player.sendMessage(Component.text("You´ve successfully merged this land!"));
                player.sendMessage(String.format("DEBUG: Connected with Land X: %d Z: %d", land.getX(), land.getZ()));

            } else {
                if (this.api.hasPlayerLand(player)) {
                    player.sendMessage(Component.text("Du besitzt bereits schon ein Land."));
                } else {
                    this.api.getLandService().createLand(landPlayer, player, playerChunk);
                    player.sendMessage(Component.text("You´ve successfully claimed this land."));
                }
            }
        });
    }


}

File: src/main/kotlin/net/onelitefeather/pandorascluster/command/commands/ClaimCommand.kt
package net.onelitefeather.pandorascluster.command.commands

import cloud.commandframework.annotations.CommandDescription
import cloud.commandframework.annotations.CommandMethod
import net.kyori.adventure.text.Component
import net.onelitefeather.pandorascluster.api.PandorasClusterApi
import net.onelitefeather.pandorascluster.extensions.ChunkUtils
import net.onelitefeather.pandorascluster.extensions.EntityUtils
import net.onelitefeather.pandorascluster.util.AVAILABLE_CHUNK_ROTATIONS
import net.onelitefeather.pandorascluster.util.IGNORE_CLAIM_LIMIT
import org.bukkit.Chunk
import org.bukkit.entity.Player

class ClaimCommand(private val pandorasClusterApi: PandorasClusterApi) : EntityUtils, ChunkUtils {

    @CommandMethod("land claim")
    @CommandDescription("Claim a free chunk")
    fun execute(player: Player) {

        val pluginPrefix = pandorasClusterApi.pluginPrefix()

        val landPlayer = pandorasClusterApi.getLandPlayer(player.uniqueId)
        if (landPlayer == null) {
            player.sendMessage(Component.translatable("player-data-not-found").arguments(
                pluginPrefix, Component.text(player.name)))
            return
        }

        val playerChunk = player.chunk
        if (pandorasClusterApi.getLandService().checkWorldGuardRegion(playerChunk)) {
            player.sendMessage(Component.translatable("worldguard-region-found").arguments(pluginPrefix))
            return
        }

        if (pandorasClusterApi.isChunkClaimed(playerChunk)) {
            player.sendMessage(Component.translatable("chunk-already-claimed").arguments(pluginPrefix))
            return
        }

        val chunkX = playerChunk.x
        val chunkZ = playerChunk.z

        pandorasClusterApi.getLandService().findConnectedChunk(player) {
            if (it != null) {

                var claimedChunk: Chunk? = null

                for(facing in AVAILABLE_CHUNK_ROTATIONS) {
                    if (claimedChunk != null) continue
                    val chunk = player.world.getChunkAt(facing.modX + chunkX, facing.modZ + chunkZ)
                    if (pandorasClusterApi.isChunkClaimed(chunk)) {
                        claimedChunk = chunk
                    }
                }

                if (claimedChunk != null) {
                    val claimedLand = pandorasClusterApi.getLand(claimedChunk)
                    if (claimedLand != null && !hasSameOwner(it, claimedLand)) {
                        player.sendMessage(Component.translatable("another-land-too-close").arguments(pluginPrefix))
                        return@findConnectedChunk
                    }
                }

                if (!it.isOwner(player.uniqueId)) {
                    player.sendMessage(Component.translatable("invalid-land-owner").arguments(pluginPrefix))
                    return@findConnectedChunk
                }

                val claimLimit = getHighestClaimLimit(player)

                // Add 1 to the current chunk count and check if the player can claim more chunks
                val newChunkCount = (pandorasClusterApi.getLandService().getChunksByLand(it) + 1)

                if(claimLimit != IGNORE_CLAIM_LIMIT && newChunkCount > claimLimit) {
                    player.sendMessage(Component.translatable("chunk.claim-limit-reached").arguments(pluginPrefix))
                    return@findConnectedChunk
                }

                this.pandorasClusterApi.getDatabaseStorageService().addChunkPlaceholder(playerChunk, it)
                player.sendMessage(Component.translatable("chunk-successfully-merged").arguments(pluginPrefix))
            } else {
                if (pandorasClusterApi.hasPlayerLand(player)) {
                    player.sendMessage(Component.translatable("player-already-has-land").arguments(pluginPrefix))
                } else {
                    pandorasClusterApi.getDatabaseStorageService().createLand(landPlayer, player, playerChunk)
                    player.sendMessage(Component.translatable("chunk-successfully-claimed").arguments(pluginPrefix))
                }
            }
        }
    }
}